src/main.rs

```diff
@@ -0,0 +1,401 @@
+                    finished_at,
+                    compute_time_sec,
+                    input: Some(task_input.task_data),
+                    proof: resp.proof,
+                    vk: None,
+                    error: resp.error,
+                }
+            }
+            Err(e) => {
+                error!("query_task method failed: {:?}", e);
+                self.build_query_task_error_response(&req, &format!("Failed to query proof: {}", e))
+            }
+        }
+    }
+}
+
+impl SnarkifyProver {
+    pub fn new(cfg: CloudProverConfig, service_id: String) -> Self {
+        let retry_wait_duration = Duration::from_secs(cfg.retry_wait_time_sec);
+        let retry_policy = ExponentialBackoff::builder()
+            .retry_bounds(retry_wait_duration / 2, retry_wait_duration)
+            .build_with_max_retries(cfg.retry_count);
+        let client = ClientBuilder::new(reqwest::Client::new())
+            .with(RetryTransientMiddleware::new_with_policy(retry_policy))
+            .build();
+
+        Self {
+            base_url: cfg.base_url,
+            api_key: cfg.api_key,
+            service_id,
+            send_timeout: Duration::from_secs(cfg.connection_timeout_sec),
+            client,
+        }
+    }
+
+    pub fn build_prove_error_response(&self, req: &ProveRequest, error_msg: &str) -> ProveResponse {
+        ProveResponse {
+            task_id: String::new(),
+            circuit_type: req.circuit_type,
+            circuit_version: req.circuit_version.clone(),
+            hard_fork_name: req.hard_fork_name.clone(),
+            status: TaskStatus::Failed,
+            created_at: 0.0,
+            started_at: None,
+            finished_at: None,
+            compute_time_sec: None,
+            input: Some(req.input.clone()),
+            proof: None,
+            vk: None,
+            error: Some(error_msg.to_string()),
+        }
+    }
+
+    pub fn build_query_task_error_response(
+        &self,
+        req: &QueryTaskRequest,
+        error_msg: &str,
+    ) -> QueryTaskResponse {
+        QueryTaskResponse {
+            task_id: req.task_id.clone(),
+            circuit_type: CircuitType::Undefined,
+            circuit_version: "".to_string(),
+            hard_fork_name: "".to_string(),
+            status: TaskStatus::Queued,
+            created_at: 0.0,
+            started_at: None,
+            finished_at: None,
+            compute_time_sec: None,
+            input: None,
+            proof: None,
+            vk: None,
+            error: Some(error_msg.to_string()),
+        }
+    }
+
+    fn build_url(&self, method: &str) -> anyhow::Result<Url> {
+        let full_url = format!("{}{}", self.base_url, method);
+        Url::parse(&full_url).map_err(|e| anyhow::anyhow!(e))
+    }
+
+    async fn get_with_token<Resp>(&self, method: &str) -> anyhow::Result<Resp>
+    where
+        Resp: serde::de::DeserializeOwned,
+    {
+        let url = self.build_url(method)?;
```

Comment:
If you get a parsing error here, you lose all the context when throwing the error up. 

You need to either fill out the context (inside anyhow macro) or add a log

Reply:
good catch!